riscv-runtime/src/hash.rs

```diff
@@ -27,3 +27,85 @@ pub fn poseidon_gl(data: [u64; 12]) -> [u64; 4] {
 
     poseidon_gl_unsafe(data)
 }
+
+/// Calls the keccakf machine
+/// Return value is placed in the output array.
+/// This is unsafe because it does not check if the u64 elements fit the Goldilocks field.
+pub fn keccakf_unsafe(input: &[u64; 25], output: &mut [u64; 25]) {
+    unsafe {
+        // syscall inputs: memory pointer to input array and memory pointer to output array
+        asm!("ecall", in("a0") input as *const [u64; 25], in("a1") output as *mut [u64; 25], in("t0") u32::from(Syscall::KeccakF));
+    }
+}
+
+/// Calls the keccakf machine
+/// This function will panic if any of the u64 elements doesn't fit the Goldilocks field.
+pub fn keccakf(input: &[u64; 25], output: &mut [u64; 25]) {
+    for &n in input.iter() {
+        assert!(n < GOLDILOCKS);
+    }
+    keccakf_unsafe(input, output)
+}
+
+// Output number of bytes for keccak-256 (32 bytes)
+const W: usize = 32;
+
+/// Keccak function that calls the keccakf machine
+/// Input is a byte array of arbitrary length and a delimiter byte
+/// Output is a byte array of length W
+pub fn keccak(data: &[u8], delim: u8) -> [u8; W] {
+    let mut b_input = [0u8; 200];
+    let mut b_output = [0u64; 25];
+    let rate = 200 - (2 * W);
+    let mut pt = 0;
+
+    // update
+    for i in 0..data.len() {
```

Comment:
Question to Rust / compiler experts: Would the compiler be smart enough here to know that it can do the XORs at the word level, not byte? Or would we be 4x more efficient by casting `data` to `&[usize]`?

Reply:
Discussed and would be out of scope of this PR for now.